infor-design/enterprise
# 1658 - Improve deprecation process

**Explain the _details_ for making this change. What existing problem does the pull request solve?**
This PR adds a couple utilities to IDS that implement a better deprecation process:
- A browser-based utility that will log warnings to the console if a deprecated method or property is used.
- An optional build-time plugin for Rollup that analyzes the source code for deprecated methods/properties
- Some component documentation has been modified to remove `private` on some public-facing, deprecated methods, to actually allow the build process to identify them.  In some cases, some much older, deprecated methods were completely removed.
- Documentation on our deprecation process, along with recommendations for use.

Additionally, this PR upgrades Rollup to `v1.6.0`, along with upgrades to some of its plugins.

**Related github/jira issue (required)**:
Closes #1658

**Steps necessary to review your pull request (required)**:
_before testing:_
- Pull this branch
- run `rm -rf node_modules && npm i` in the terminal
- Review the documentation in (/docs/CODING_STANDARDS.md#Deprecations)

_test the compile-time plugin:_
- Run `npx rollup -c --verbose` to see Rollup run with a list of deprecated API methods/properties.
- Run `npm run build`.  Ensure the entire build completes.  Rollup should not identify any deprecations in the console.
- Run `npm run build -- --verbose`.  Ensure the entire build completes.  This time, Rollup should output a list of deprecated API methods/properties, among other things.

_test the browser notifications:_
- Run the demoapp
- Open a browser to http://localhost:4000/components/blockgrid/example-paging.html
- Open a Dev Tools console.
- Click on any of the blocks in the Blockgrid.  A warning in the dev tools about using the deprecated `selectBlock()` method should appear.  This is because the Blockgrid is using the original method name instead of the new method name internally (this is noted in #1820).
- Replicate this by opening a sample page for any of the Components listed with deprecated API methods/properties in the build output.

![Screen Shot 2019-03-19 at 1 01 47 PM](https://user-images.githubusercontent.com/3249601/54626351-5fafc900-4a47-11e9-9860-86fd2f841a4e.png)
